### PR TITLE
Add a simple test case for thread_manager.

### PR DESCRIPTION
**Pull request by egnor**
Refactor the fixtures in conftest into a single fixture class (DiscordMock) that provides the same functions a bit more conveniently (without having to request them individually) and more importantly has functions to simulate various changes and generate the appropriate notification events.

Use that to make a super basic test case of thread channel creation.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -10,122 +11,235 @@
...
         message.guild = guild
         message.channel = channel
         message.author = author
         message.content = content
+        message.reactions = []
         return message
 
-    return _factory
+    #
+    # Helper methods to update data in useful ways.
```
Comment (anaulin) [inline tests/conftest.py]: This comment makes me wonder if you are planning to add helpers to update data in useless ways. 😹

File: tests/conftest.py [hunk 2]
```diff
@@ -10,122 +11,235 @@
...
+        guild.channels.append(channel)
+        self.queue_event('on_guild_channel_create', channel)
+        return channel
+
+
+@pytest.fixture
+def discord_mock(mocker, event_loop):
+    """Fixture class to generate Mocks for Discord client library objects."""
+
+    yield DiscordMockFixture(mocker)  # Keep event loop until teardown.
```
Comment (anaulin) [inline tests/conftest.py:244]: Neat pattern.

File: tests/conftest.py [hunk 3]
```diff
@@ -10,122 +11,235 @@
...
+
+        while self.event_queue:
+            batch, self.event_queue = self.event_queue, []
+            for event_name, args, kwargs in batch:
+                handler = getattr(self.context.client, event_name, None)
+                if handler is not None:
+                    await handler(*args, **kwargs)
+
+    def sim_reaction(self, message, unicode, user, delta):
+        """Simulates an emoji reaction changes and queues notification events.
```
Comment (anaulin) [inline tests/conftest.py]: s/changeS/change/ (or something else about the concordance is discordant)

File: tests/conftest.py [hunk 4]
```diff
@@ -10,122 +11,235 @@
...
+
+        Args:
+            message - the message object to modify
+            unicode - unicode of emoji to add/remove
+            user - the user adding/removing the emoji
+            delta - +1 to add, -1 to remove
+        """
+
+        if delta not in (-1, +1):
+            raise ValueError('reaction {unicode} delta {delta} not -1 or +1')
```
Comment (anaulin) [inline tests/conftest.py]: Is this meant to be an f-string? (same for like 211 below)

File: tests/conftest.py [hunk 5]
```diff
@@ -10,122 +11,235 @@
...
+            spec=discord.RawReactionActionEvent, name='raw_reaction_event')
+        event.message_id = message.id
+        event.user_id = user.id
+        event.channel_id = message.channel.id
+        event.guild_id = message.guild.id
+        event.emoji = message_reaction.emoji
+        event.member = user
+        event.event_type = f'REACTION_{"ADD" if delta > 0 else "REMOVE"}'
+        self.queue_event(
+            f'on_raw_reaction_{"add" if delta > 0 else "remove"}', event)
```
Comment (anaulin) [inline tests/conftest.py]: I find these interpolated if-elses with double-quoted strings kinda hard to read, and also a little redundant. Possible alternative:
```python
action = "ADD" if delta > 0 else "REMOVE"
event.event_type = f'REACTION_{action}'
self.queue_event(f'on_raw_reaction_{action.lower()}', event)
```
("action" might not be the best name for this, but you get the gist)

**Review by anaulin — APPROVED** [review]
Very nice. 🥇  🧵  🧪 

Minor (aka non-blocking) comments below.

File: tests/conftest.py (hunk 5 above)
Comment (dcsan) [reply tests/conftest.py]: the basic way isn't a lot longer, and easier for later someone else to read:

```py
reaction = "REACTION_ADD" if delta > 0 else "REACTION_REMOVE"
```

BDFL really had a thing against ternary eh. also i used to like ruby's `unless`

File: tests/conftest.py (hunk 1 above)
Comment (egnor) [reply tests/conftest.py]: Oh boy am I!

File: tests/conftest.py (hunk 3 above)
Comment (egnor) [reply tests/conftest.py]: Done.

File: tests/conftest.py (hunk 4 above)
Comment (egnor) [reply tests/conftest.py]: Good catch!

File: tests/conftest.py (hunk 5 above)
Comment (egnor) [reply tests/conftest.py]: Simplified!

**Pull request merged**